test/apps/iodemo/io_demo.cc

```diff
@@ -1309,8 +1310,13 @@ class DemoServer : public P2pDemoCommon {
         w->init(this, conn, msg->sn, msg->conn_id, iov,
                 &conn_stat.completions<IO_WRITE>());
         conn_stat.bytes<IO_WRITE>() += msg->data_size;
-        conn->recv_am_data((*iov)[0].buffer(), (*iov)[0].size(),
-                           (*iov)[0].memh(), data_desc, w);
+        if (!ucx_am_is_rndv(data_desc)) {
+            conn->recv_am_data(NULL, 0,
+                               NULL, data_desc, w);
+        } else {
+            conn->recv_am_data((*iov)[0].buffer(), (*iov)[0].size(),
+                               (*iov)[0].memh(), data_desc, w);
+        }
```

Comment:
I restore it because I got a SIGSEGV when (!ucx_am_is_rndv(data_desc)). i believe (*iov)[0] caused it.

Reply:
can you pls check the reason for this segfault? and print a backtrace?
also, can you pls add a stage to https://github.com/openucx/ucx/blob/6b404f6984cd864d9d245328820a13b04f85c2f5/buildlib/pr/io_demo/io-demo.yml#L44, to run active messages on ConnectX-6 with data check (add -q to args) so this case would run in CI?

Reply:
i believe (*iov)[0] caused SIGSEGV because iov was empty when (!ucx_am_is_rndv(data_desc)), here is the bt:
Caught signal 11 (Segmentation fault: address not mapped to object at address 0x30)
==== backtrace (tid: 671652) ====
 0 0x0000000000012b20 .annobin_sigaction.c()  sigaction.c:0
 1 0x000000000041b428 DemoServer::dispatch_am_message()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/io_demo.cc:665
 2 0x000000000041b428 DemoServer::dispatch_am_message()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/io_demo.cc:1373
 3 0x00000000004094e7 UcxContext::am_recv_callback()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/ucx_wrapper.cc:696
 4 0x000000000002445a ucp_am_invoke_cb()  /.autodirect/swgwork/huaxiangf/o-ucx/src/ucp/core/ucp_am.c:1203
 5 0x000000000002445a ucp_am_handler_reply()  /.autodirect/swgwork/huaxiangf/o-ucx/src/ucp/core/ucp_am.c:1310
 6 0x0000000000036e3c uct_iface_invoke_am()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/base/uct_iface.h:773
 7 0x0000000000036e3c uct_rc_mlx5_iface_common_am_handler()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/ib/rc/accel/rc_mlx5.inl:427
 8 0x0000000000036e3c uct_rc_mlx5_iface_common_poll_rx()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/ib/rc/accel/rc_mlx5.inl:1455
 9 0x0000000000036e3c uct_rc_mlx5_iface_progress()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/ib/rc/accel/rc_mlx5_iface.c:169
10 0x0000000000036e3c uct_rc_mlx5_iface_progress_cyclic()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/ib/rc/accel/rc_mlx5_iface.c:179
11 0x000000000003bde2 ucs_callbackq_dispatch()  /.autodirect/swgwork/huaxiangf/o-ucx/src/ucs/datastruct/callbackq.h:211
12 0x000000000003bde2 uct_worker_progress()  /.autodirect/swgwork/huaxiangf/o-ucx/src/uct/api/uct.h:2600
13 0x000000000003bde2 ucp_worker_progress()  /.autodirect/swgwork/huaxiangf/o-ucx/src/ucp/core/ucp_worker.c:2665
14 0x0000000000406282 UcxContext::progress_worker_event()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/ucx_wrapper.cc:388
15 0x000000000040bbd5 UcxContext::progress()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/ucx_wrapper.cc:225
16 0x000000000040fc1e DemoServer::run()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/io_demo.cc:1222
17 0x000000000040568f do_server()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/io_demo.cc:2853
18 0x000000000040568f main()  /.autodirect/swgwork/huaxiangf/o-ucx/test/apps/iodemo/io_demo.cc:2918
19 0x0000000000023493 __libc_start_main()  ???:0
20 0x00000000004059fe _start()  ???:0
Will add -q to CI.

Reply:
I see - iov is empty and it has only non-NULL extra_buf in case of eager.
So, can we move this code outside of recv_am_data() to avoid extra branch:
```c
    if (!_context.ucx_am_is_rndv(data_desc)) {
        (*callback)(UCS_OK);
        return true;
    }
```

so it will be something like:
```c
        if (!ucx_am_is_rndv(data_desc)) {
            (*w)(UCS_OK);
        } else {
            conn->recv_am_data((*iov)[0].buffer(), (*iov)[0].size(),
                               (*iov)[0].memh(), data_desc, w);
        }
```

@brminich WDYT?

Reply:
sounds good

Reply:
I got [static check errors](https://dev.azure.com/ucfconsort/ucx/_build/results?buildId=34525&view=logs&j=e646a2ae-9139-5fa9-0995-85819d80e271&t=40ba7047-ddc0-522c-15d2-d7fcc7934f9b&l=1361) after applying the code changes suggested as above:
static checkers found errors:
Error: CLANG_WARNING:
/test/apps/iodemo/io_demo.cc:1319:5: warning: Potential leak of memory pointed to by 'w' [cplusplus.NewDeleteLeaks]

Reply:
ok let's keep it as is and just add CI with data validation as described in https://github.com/openucx/ucx/pull/7827/files#r778730798

Reply:
done